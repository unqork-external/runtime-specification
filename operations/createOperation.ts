import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
