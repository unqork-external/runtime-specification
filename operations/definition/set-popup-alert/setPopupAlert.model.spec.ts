/* eslint-disable max-len */
import { SetPopupAlertOperation } from './setPopupAlert.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('Set Property Model', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(SetPopupAlertOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toStrictEqual({
      definitions: {
        PopUpAlertOptions: {
          properties: {
            allowOutsideClick: {
              description: 'Toggle the ability to close popup with a click outside of the popup modal',
              type: 'boolean',
            },
            apiCallErrors: {
              description: 'Errors we recieve if the api call fails',
              type: 'object',
            },
            text: {
              description: 'A description for the popup.',
              type: 'string',
            },
            title: {
              description: 'The title of the popup, as HTML.',
              type: 'string',
            },
            buttonsStyling: {
              description:
                'Apply default styling to buttons. If you want to use your own classes, set this parameter to false',
              type: 'boolean',
            },
            cancelButtonText: {
              default: 'Cancel',
              description: 'Use this to change the text on the "Cancel" button.',
              type: 'string',
            },
            confirmButtonText: {
              default: 'OK',
              description: 'Use this to change the text on the "Confirm" button.',
              type: 'string',
            },
            confirmButtonColor: {
              description: 'Use this to change the color on the "Confirm" button.',
              type: 'string',
            },
            customClass: {
              description: 'Used to add classes for components inside of the modal.',
              type: 'object',
            },
            shouldAutoClose: {
              default: false,
              description: 'Should autoclose the popup without user interaction',
              type: 'boolean',
            },
            icon: {
              description: 'An icon for the modal. SweetAlert comes with 5 built-in icons that you can use',
              type: 'string',
            },
            showCancelButton: {
              description:
                'If set to true, a "Cancel" button will be shown, which the user can click on to dismiss the modal.',
              type: 'boolean',
            },
            showConfirmButton: {
              description:
                'If set to true, a "Confirm" button will be shown, which the user can click on to dismiss the modal.',
              type: 'boolean',
            },
            timer: {
              description: 'Amount of time before the popup modal automatically closes',
              type: 'number',
            },
            validationErrorCount: {
              description: 'Total errors from our validation system',
              type: 'number',
            },
          },
          type: 'object',
        },
        SetPopupAlertOptions: {
          properties: {
            value: {
              $ref: '#/definitions/PopUpAlertOptions',
            },
            targetKey: {
              description: 'The key or path to the container we intend to target for an operation',
              type: 'string',
            },
          },
          required: ['value'],
          type: 'object',
        },
      },
      properties: {
        options: {
          $ref: '#/definitions/SetPopupAlertOptions',
        },
        type: {
          const: 'SET_POPUP_ALERT',
          examples: ['SET_POPUP_ALERT'],
          type: 'string',
        },
        name: {
          description: 'Name of the operation',
          type: 'string',
        },
      },
      required: ['options'],
      type: 'object',
    })
  })

  it('should validate appropriately', () => {
    const op = createOperation(OperationTypes.SET_POPUP_ALERT, {
      value: {
        buttonsStyling: true,
        cancelButtonText: 'Cancel',
        confirmButtonText: 'OK',
        icon: 'warning',
        showCancelButton: true,
        text: 'Please be sure',
        title: 'Are you sure?',
      },
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should property set up the popup alert value', () => {
    const op = createOperation(OperationTypes.SET_POPUP_ALERT, {
      value: {
        buttonsStyling: true,
        cancelButtonText: 'Cancel',
        confirmButtonText: 'OK',
        icon: 'warning',
        showCancelButton: true,
        text: 'Please be sure',
        title: 'Are you sure?',
      },
    })

    expect(op.options.value.buttonsStyling).toStrictEqual(true)
    expect(op.options.value.cancelButtonText).toStrictEqual('Cancel')
    expect(op.options.value.confirmButtonText).toStrictEqual('OK')
    expect(op.options.value.icon).toStrictEqual('warning')
    expect(op.options.value.showCancelButton).toStrictEqual(true)
    expect(op.options.value.text).toStrictEqual('Please be sure')
    expect(op.options.value.title).toStrictEqual('Are you sure?')
  })
})
