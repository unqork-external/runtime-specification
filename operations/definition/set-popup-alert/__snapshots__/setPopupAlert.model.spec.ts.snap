// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set Property Model should match this JSON Schema 1`] = `
{
  "autogenDisabled": true,
  "definitions": {
    "PopUpAlertOptions": {
      "properties": {
        "allowOutsideClick": {
          "description": "Toggle the ability to close popup with a click outside of the popup modal",
          "type": "boolean",
        },
        "apiCallErrors": {
          "description": "Errors we recieve if the api call fails",
          "type": "object",
        },
        "buttonsStyling": {
          "description": "Apply default styling to buttons. If you want to use your own classes, set this parameter to false",
          "type": "boolean",
        },
        "cancelButtonText": {
          "default": "Cancel",
          "description": "Use this to change the text on the "Cancel" button.",
          "type": "string",
        },
        "confirmButtonColor": {
          "description": "Use this to change the color on the "Confirm" button.",
          "type": "string",
        },
        "confirmButtonText": {
          "default": "OK",
          "description": "Use this to change the text on the "Confirm" button.",
          "type": "string",
        },
        "customClass": {
          "description": "Used to add classes for components inside of the modal.",
          "type": "object",
        },
        "didOpen": {
          "description": "We can provide a function to SweetAlert to asynchronously run after the popup has been shown on screen. Provides popup DOM element as the argument. In previous versions of SweetAlert, this hook was named onOpen.",
          "type": "object",
        },
        "icon": {
          "description": "An icon for the modal. SweetAlert comes with 5 built-in icons that you can use",
          "type": "string",
        },
        "shouldAutoClose": {
          "default": false,
          "description": "Should autoclose the popup without user interaction",
          "type": "boolean",
        },
        "showCancelButton": {
          "description": "If set to true, a "Cancel" button will be shown, which the user can click on to dismiss the modal.",
          "type": "boolean",
        },
        "showConfirmButton": {
          "description": "If set to true, a "Confirm" button will be shown, which the user can click on to dismiss the modal.",
          "type": "boolean",
        },
        "text": {
          "description": "A description for the popup.",
          "type": "string",
        },
        "timer": {
          "description": "Amount of time before the popup modal automatically closes",
          "type": "number",
        },
        "title": {
          "description": "The title of the popup, as HTML.",
          "type": "string",
        },
        "validationErrorCount": {
          "description": "Total errors from our validation system",
          "type": "number",
        },
      },
      "type": "object",
    },
    "SetPopupAlertOptions": {
      "properties": {
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "value": {
          "$ref": "#/definitions/PopUpAlertOptions",
        },
      },
      "required": [
        "value",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/SetPopupAlertOptions",
    },
    "type": {
      "const": "SET_POPUP_ALERT",
      "examples": [
        "SET_POPUP_ALERT",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
